src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -2791,6 +2792,67 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+        {
+            // If the assignment target is a tuple, we want to decompose the tuple and push each element back onto the stack, so that if the value
+            // has control flow the individual elements are captured. Then we can recompose the tuple after operation.Value has been visited.
+            // We do this to keep the graph sane, so that users don't have to track a tuple captured via flow control when it's not really
+            // the tuple that's been captured, it's the operands to the tuple.
+            pushTargetAndUnwrapTupleIfNecessary(operation.Target);
+            IOperation visitedValue = Visit(operation.Value);
+            IOperation visitedTarget = popTargetAndWrapTupleIfNecessary(operation.Target);
+
+            return new DeconstructionAssignmentExpression(visitedTarget, visitedValue, semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);
+
+            // Recursively push nexted values onto the stack for visiting
+            void pushTargetAndUnwrapTupleIfNecessary(IOperation value)
+            {
+                if (value.Kind == OperationKind.Tuple)
+                {
+                    var tuple = (ITupleOperation)value;
+
+                    foreach (IOperation element in tuple.Elements)
+                    {
+                        pushTargetAndUnwrapTupleIfNecessary(element);
+                    }
+                }
+                else
+                {
+                    _evalStack.Push(Visit(value));
+                }
+            }
+
+            // Recursively pop nested tuple values off the stack after visiting
+            IOperation popTargetAndWrapTupleIfNecessary(IOperation value)
+            {
+                if (value.Kind == OperationKind.Tuple)
+                {
+                    var tuple = (ITupleOperation)value;
+                    var elementBuilder = ArrayBuilder<IOperation>.GetInstance();
+                    for (int i = tuple.Elements.Length - 1; i >= 0; i--)
+                    {
+                        elementBuilder.Add(popTargetAndWrapTupleIfNecessary(tuple.Elements[i]));
+                    }
+                    elementBuilder.ReverseContents();
+                    return new TupleExpression(elementBuilder.ToImmutableAndFree(), semanticModel: null, tuple.Syntax, tuple.Type, tuple.NaturalType, tuple.ConstantValue, tuple.IsImplicit);
+                }
+                else
+                {
+                    return _evalStack.Pop();
+                }
+            }
+        }
+
+        public override IOperation VisitDeclarationExpression(IDeclarationExpressionOperation operation, int? captureIdForResult)
+        {
+            return new DeclarationExpression(Visit(operation.Expression), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);
```

Comment:
>return new DeclarationExpression(Visit(operation.Expression), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit); [](start = 12, length = 164)

Consider reusing logic from ```VisitDeconstructionAssignment``` if ```operation.Expression``` is a tuple. I think this will make implementation more robust. We want to guarantee that the shape of the structure is not changed no matter what.

Reply:
I'll do this in a followup PR, filed [https://github.com/dotnet/roslyn/issues/26151](https://github.com/dotnet/roslyn/issues/26151) to track.

---
In reply to: [181504915](https://github.com/dotnet/roslyn/pull/26108#discussion_r181504915) [](ancestors = 181504915)